src/lib/ngsiNotify/Notifier.cpp

```diff
@@ -50,32 +50,302 @@ Notifier::~Notifier (void)
+                   paramP->httpInfo,
+                   paramP->tenant,
+                   paramP->xauthToken,
+                   paramP->fiwareCorrelator,
+                   paramP->renderFormat,
+                   paramP->attrsOrder);
+  }
+
+  delete paramP;
+  return NULL;
+}
+
+
+
 /* ****************************************************************************
 *
 * Notifier::sendNotifyContextRequest -
 */
 void Notifier::sendNotifyContextRequest
 (
-  NotifyContextRequest*            ncr,
-  const std::string&               url,
+  NotifyContextRequest*            ncrP,
+  const ngsiv2::HttpInfo&          httpInfo,
   const std::string&               tenant,
   const std::string&               xauthToken,
   const std::string&               fiwareCorrelator,
   RenderFormat                     renderFormat,
   const std::vector<std::string>&  attrsOrder
 )
 {
-    ConnectionInfo ci;
+    ConnectionInfo  ci;
+    Verb            verb = httpInfo.verb;
+
+    LM_W(("KZ: In sendNotifyContextRequest"));
+    if (verb == NOVERB)
+    {
+      // Default verb/method is POST
+      verb = POST;
+    }
+
+    //
+    // If any of the 'template parameters' is used by the subscripion, then it is not a question of an ordinary notification but one using templates.
+    // Ordinary notifications are simply sent, all of it as one message.
+    //
+    // Notifications for subscriptions using templates are more complex:
+    //   - if there is more than one entity for the notification, split into N notifications, one per entity
+    //   - if 'url' contains substitution keys, substitute the keys for their current values
+    //   - if 'qs' contains query strings, add this to 'url', with the proper substitutions done
+    //   - if 'headers' is non-empty, perform eventual substitutions and make sure the information is added as HTTP headers for the notification
+    //   - if 'payload' is given, use that string as template instead of the default payload string, substituting all fields that are to be substituted
+    //
+    // Redirect to the method sendNotifyContextRequestAsPerTemplate() when 'httpInfo.extended' is TRUE.
+    // 'httpInfo.extended' is FALSE by default and set to TRUE by the json parser.
+    //
+    if (httpInfo.extended)
+    {
+      LM_W(("KZ: Calling sendNotifyContextRequestAsPerTemplate"));
+
+      NotificationAsTemplateParams* paramP = new NotificationAsTemplateParams();
+
+      paramP->ncrP             = ncrP->clone();
+      paramP->httpInfo         = httpInfo;
+      paramP->tenant           = tenant;
+      paramP->xauthToken       = xauthToken;
+      paramP->fiwareCorrelator = fiwareCorrelator;
+      paramP->renderFormat     = renderFormat;
+      paramP->attrsOrder       = attrsOrder;
+
+#if 0
+      sendNotifyContextRequestAsPerTemplate(paramP);
+#else
+      pthread_t  tid;
+      int        r = pthread_create(&tid, NULL, sendNotifyContextRequestAsPerTemplate, (void*) paramP);
```

Comment:
What is the purpose of creating a new thread at this point?

I mean, it could make sense to create one thread per HTTP request in the for loop inside sendNotifyContextRequestAsPerTemplate(), so each notification could be send in a paralel thread. However, I don't see the point of doing it at this point...